File: gren/kit/graphs/reducers/run-command.py
# Deterministic command runner (tests, builds, linters). Not a side effect: it only reads/executes and reports.
# input: { cwd?, command?, args? }  args: { command: "python", args: ["-m","pytest","-q"], timeout_ms: 600000, cwd?: ".", ok_pattern?: "regex" }
import os
import re
import shutil
import subprocess
import time


def _tail(s: str, n: int = 4000) -> str:
    return f"…{s[-n:]}" if len(s) > n else s


def reduce(input, args, ctx):
    command = str(input.get("command") or args.get("command") or "python")
    cmd_args = [str(a) for a in (input.get("args") if isinstance(input.get("args"), list) else args.get("args") if isinstance(args.get("args"), list) else [])]
    cwd = os.path.abspath(str(input.get("cwd") or args.get("cwd") or "."))
    timeout = float(args.get("timeout_ms", 600000)) / 1000
    started = time.time()
    ctx.log(f"run-command: {command} {' '.join(cmd_args)} (cwd {cwd})")
    exe = shutil.which(command) or command
    env = {**os.environ, "CI": "1", "FORCE_COLOR": "0", "PYTHONIOENCODING": "utf-8"}
    try:
        p = subprocess.run([exe, *cmd_args], cwd=cwd, env=env, capture_output=True, text=True, encoding="utf-8", errors="replace", timeout=timeout)
        out, err, code, timed_out = p.stdout, p.stderr, p.returncode, False
    except subprocess.TimeoutExpired as e:
        out, err, code, timed_out = (e.stdout or b"").decode("utf-8", "replace") if isinstance(e.stdout, bytes) else (e.stdout or ""), (e.stderr or b"").decode("utf-8", "replace") if isinstance(e.stderr, bytes) else (e.stderr or ""), -1, True
    except OSError as e:
        out, err, code, timed_out = "", str(e), -1, False
    ok_pattern = args.get("ok_pattern")
    passed = code == 0 and (not ok_pattern or re.search(str(ok_pattern), out + err) is not None)
    return {
        "command": f"{command} {' '.join(cmd_args)}".strip(), "cwd": cwd.replace("\\", "/"), "exit_code": code, "passed": passed, "timed_out": timed_out,
        "duration_ms": int((time.time() - started) * 1000), "stdout_tail": _tail(out), "stderr_tail": _tail(err),
    }

File: gren/v1/kit/graphs/reducers/run-command.js
// Deterministic command runner (tests, builds, linters). Not a side effect: it only reads/executes and reports.
// input: { cwd?, command?, args? }  args: { command: "npm", args: ["test"], timeout_ms: 600000, cwd?: ".", ok_pattern?: "regex" }
import { spawn } from "node:child_process";
import path from "node:path";

export default async function runCommand(input, args, ctx) {
  const command = String(input.command ?? args.command ?? "npm");
  const cmdArgs = Array.isArray(input.args) ? input.args.map(String) : Array.isArray(args.args) ? args.args.map(String) : [];
  const cwd = path.resolve(String(input.cwd ?? args.cwd ?? "."));
  const timeout = Number(args.timeout_ms ?? 600000);
  const started = Date.now();
  ctx.log(`run-command: ${command} ${cmdArgs.join(" ")} (cwd ${cwd})`);
  const result = await new Promise((resolve) => {
    const child = spawn(command, cmdArgs, { cwd, shell: process.platform === "win32", env: { ...process.env, CI: "1", FORCE_COLOR: "0" } });
    let out = "", err = "";
    child.stdout.on("data", (d) => (out += d.toString()));
    child.stderr.on("data", (d) => (err += d.toString()));
    const t = setTimeout(() => { child.kill(); resolve({ exit_code: -1, timed_out: true, stdout: out, stderr: err }); }, timeout);
    child.on("close", (code) => { clearTimeout(t); resolve({ exit_code: code ?? -1, timed_out: false, stdout: out, stderr: err }); });
    child.on("error", (e) => { clearTimeout(t); resolve({ exit_code: -1, timed_out: false, stdout: out, stderr: String(e.message) }); });
  });
  const tail = (s, n = 4000) => (s.length > n ? `…${s.slice(-n)}` : s);
  const okPattern = args.ok_pattern ? new RegExp(String(args.ok_pattern)) : null;
  const passed = result.exit_code === 0 && (!okPattern || okPattern.test(result.stdout + result.stderr));
  return {
    command: `${command} ${cmdArgs.join(" ")}`.trim(),
    cwd: cwd.replace(/\\/g, "/"),
    exit_code: result.exit_code,
    passed,
    timed_out: result.timed_out,
    duration_ms: Date.now() - started,
    stdout_tail: tail(result.stdout),
    stderr_tail: tail(result.stderr),
  };
}
